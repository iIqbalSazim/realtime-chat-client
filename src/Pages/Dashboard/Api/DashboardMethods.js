import { getRequest, postRequest } from "../../../Config/Axios/AxiosConfig";

export const fetchAllUsers = async () => {
  return await getRequest("users");
};

export const fetchUserMessageHistory = async (receiverId) => {
  return await getRequest(`messages?chatting_with=${receiverId}`);
};

export const sendMessage = async (message) => {
  return await postRequest("messages", message);
};
