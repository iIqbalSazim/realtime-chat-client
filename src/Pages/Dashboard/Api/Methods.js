import { getRequest } from "../../../Config/Axios/AxiosConfig";

export const fetchAllUsers = async () => {
  return await getRequest("users");
};

export const fetchUserMessageHistory = async (senderId, receiverId) => {
  return await getRequest(
    `messages?sender_id=${senderId}&receiver_id=${receiverId}`
  );
};
