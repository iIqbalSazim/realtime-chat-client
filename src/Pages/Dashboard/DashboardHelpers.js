export const convertApiMessagesToChatMessages = (
  apiMessages,
  currentUser,
  chattingWith
) => {
  return apiMessages.map((message) => {
    if (message.sender_id === currentUser.id) {
      return {
        id: message.id,
        message: message.content,
        sender: currentUser.username,
        receiver: chattingWith.username,
        sender_id: currentUser.id,
        receiver_id: chattingWith.id,
        direction: "outgoing",
        position: "single",
      };
    } else {
      return {
        id: message.id,
        message: message.content,
        sender: chattingWith.username,
        receiver: currentUser.username,
        sender_id: chattingWith.id,
        receiver_id: currentUser.id,
        direction: "incoming",
        position: "single",
      };
    }
  });
};

export const convertApiMessageToChatMessage = (
  message,
  currentUser,
  chattingWith
) => {
  if (message.sender_id === currentUser.id) {
    return {
      id: message.id,
      message: message.content,
      sender: currentUser.username,
      receiver: chattingWith.username,
      sender_id: currentUser.id,
      receiver_id: chattingWith.id,
      direction: "outgoing",
      position: "single",
    };
  } else {
    return {
      id: message.id,
      message: message.content,
      sender: chattingWith.username,
      receiver: currentUser.username,
      sender_id: chattingWith.id,
      receiver_id: currentUser.id,
      direction: "incoming",
      position: "single",
    };
  }
};
