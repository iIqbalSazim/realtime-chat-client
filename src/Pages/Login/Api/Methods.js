import { postRequest } from "../../../Config/Axios/AxiosConfig";

export const loginUser = async (user) => {
  return await postRequest("login", user);
};

export const revokeToken = async (token) => {
  return await postRequest("revoke_token", token);
};
