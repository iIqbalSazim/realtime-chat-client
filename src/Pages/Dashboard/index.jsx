/* eslint-disable react-hooks/exhaustive-deps */
import { useEffect, useState } from "react";
import { useSelector } from "react-redux";
import { MainContainer, Sidebar } from "@chatscope/chat-ui-kit-react";

import Navbar from "../../Shared/Components/Navbar";

import Chat from "./Chat/Chat";
import Conversations from "./Conversations/Conversations";

import { fetchAllUsers } from "./Api/Methods";

const Dashboard = () => {
  const [users, setUsers] = useState([]);

  const currentUser = useSelector((state) => state.auth.user);
  const chattingWith = useSelector((state) => state.chat.user);

  const getAllUsers = async () => {
    const response = await fetchAllUsers();

    const allUsers = response.data;
    const filteredUsers = allUsers.filter((user) => user.id !== currentUser.id);

    setUsers(filteredUsers);
  };

  useEffect(() => {
    getAllUsers();
  }, []);

  return (
    <main className="relative pr-[1px] h-full">
      <Navbar />
      <MainContainer className="h-screen min-h-[700px]">
        <Sidebar position="left" scrollable={false}>
          <div className="flex justify-center items-center bg-slate-200 h-20">
            <p className="uppercase">{currentUser.username}</p>
          </div>

          <Conversations users={users} />
        </Sidebar>

        {chattingWith ? (
          <Chat chattingWith={chattingWith} currentUser={currentUser} />
        ) : (
          <div className="w-full items-center flex justify-center">
            <div>Select someone to chat with</div>
          </div>
        )}
      </MainContainer>
    </main>
  );
};

export default Dashboard;
