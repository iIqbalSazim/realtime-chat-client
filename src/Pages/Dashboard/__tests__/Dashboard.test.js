import { render, screen, waitFor } from "@testing-library/react";
import { act } from "react-test-renderer";
import configureStore from "redux-mock-store";
import { Provider } from "react-redux";
import { MemoryRouter } from "react-router-dom";
import ActionCable from "actioncable";

import * as DashboardMethods from "../Api/DashboardMethods";
import Dashboard from "../Dashboard";
import {
  mockChattingWith,
  mockCurrentUser,
  mockUsers,
  newUser,
} from "../../../MockData/userItems";

const mockStore = configureStore([]);

describe("Dashboard", () => {
  afterEach(() => {
    jest.resetAllMocks();
  });

  test("renders correctly and fetches all users", async () => {
    const initialState = {
      auth: { user: mockCurrentUser },
      chat: { chatUser: null },
    };

    const store = mockStore(initialState);

    jest.spyOn(DashboardMethods, "fetchAllUsers").mockResolvedValueOnce({
      data: [...mockUsers],
    });

    const { container } = render(
      <Provider store={store}>
        <MemoryRouter>
          <Dashboard />
        </MemoryRouter>
      </Provider>
    );
    const userNameElement = screen.getByText(mockCurrentUser.username);
    const selectSomeoneToChatWithElement = screen.getByText(
      "Select someone to chat with"
    );

    await waitFor(() =>
      expect(selectSomeoneToChatWithElement).toBeInTheDocument()
    );
    expect(userNameElement).toBeInTheDocument();
    expect(DashboardMethods.fetchAllUsers).toHaveBeenCalled();

    expect(container).toMatchSnapshot();
  });

  test("does not render 'Select someone to chat with' when there is an active chat", async () => {
    const initialState = {
      auth: { user: mockCurrentUser },
      chat: { chatUser: mockChattingWith, messages: [] },
    };

    const store = mockStore(initialState);

    jest.spyOn(DashboardMethods, "fetchAllUsers").mockResolvedValueOnce({
      data: [],
    });

    const { container } = render(
      <Provider store={store}>
        <MemoryRouter>
          <Dashboard />
        </MemoryRouter>
      </Provider>
    );

    await waitFor(() =>
      expect(
        screen.queryByText("Select someone to chat with")
      ).not.toBeInTheDocument()
    );

    expect(container).toMatchSnapshot();
  });
});

describe("ActionCable Socket", () => {
  test("establishes connection and invokes received method with newUser", async () => {
    const initialState = {
      auth: { user: mockCurrentUser },
      chat: { chatUser: mockChattingWith, messages: [] },
    };

    const store = mockStore(initialState);

    jest.spyOn(DashboardMethods, "fetchAllUsers").mockResolvedValueOnce({
      data: [],
    });

    const mockReceived = jest.fn();
    const mockUnsubscribe = jest.fn();

    const mockSubscription = {
      create: jest.fn().mockReturnValue({
        received: mockReceived,
        unsubscribe: mockUnsubscribe,
      }),
    };

    const createConsumerSpy = jest.spyOn(ActionCable, "createConsumer");

    createConsumerSpy.mockReturnValueOnce({
      subscriptions: {
        ...mockSubscription,
      },
    });

    render(
      <Provider store={store}>
        <MemoryRouter>
          <Dashboard />
        </MemoryRouter>
      </Provider>
    );

    await waitFor(() => {
      expect(createConsumerSpy).toHaveBeenCalled();

      expect(mockSubscription.create).toHaveBeenCalledWith(
        {
          channel: "AppearanceChannel",
        },
        {
          received: expect.any(Function),
        }
      );
    });

    act(() => {
      mockSubscription.create.mock.calls[0][1].received(newUser);
    });

    expect(screen.getByText(/New user/i)).toBeInTheDocument();
  });
});
