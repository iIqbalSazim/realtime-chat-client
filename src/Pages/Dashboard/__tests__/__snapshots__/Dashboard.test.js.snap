// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dashboard does not render 'Select someone to chat with' when there is an active chat 1`] = `
<div>
  <main
    class="relative pr-[1px] h-full"
  >
    <nav
      class="flex justify-center gap-5 p-2 absolute z-50 right-2"
    >
      <a
        class="btn btn-sm btn-secondary btn-outline"
        href="/1"
      >
        Dashboard
      </a>
      <button
        class="btn btn-sm btn-accent"
      >
        Logout
      </button>
    </nav>
    <div
      class="cs-main-container h-screen min-h-[700px]"
    >
      <div
        class="cs-sidebar cs-sidebar--left"
      >
        <div
          class="flex justify-center items-center bg-slate-200 h-20"
        >
          <p>
            User1
          </p>
        </div>
        <div>
          No users found
        </div>
      </div>
      <div
        class="cs-chat-container"
      >
        <div
          class="cs-conversation-header mb-2"
        >
          <div
            class="cs-conversation-header__content"
          >
            <div
              class="cs-conversation-header__user-name"
            >
              User2
            </div>
            <div
              class="cs-conversation-header__info"
            />
          </div>
        </div>
        <div
          class="cs-message-list"
        >
          <div
            class="scrollbar-container cs-message-list__scroll-wrapper ps"
            data-cs-message-list=""
          >
            <div
              class="items-center flex justify-center w-full my-2"
            >
              <div
                class="btn btn-neutral btn-outline btn-wide"
              >
                Say Hi!
              </div>
            </div>
            <div
              class="cs-message-list__scroll-to"
            />
            <div
              class="ps__rail-x"
              style="left: 0px; top: 0px;"
            >
              <div
                class="ps__thumb-x"
                style="left: 0px; width: 0px;"
                tabindex="0"
              />
            </div>
            <div
              class="ps__rail-y"
              style="top: 0px; left: 0px;"
            >
              <div
                class="ps__thumb-y"
                style="top: 0px; height: 0px;"
                tabindex="0"
              />
            </div>
          </div>
        </div>
        <div
          class="cs-message-input"
        >
          <div
            class="cs-message-input__content-editor-wrapper"
          >
            <div
              class="scrollbar-container cs-message-input__content-editor-container ps"
            >
              <div
                class="cs-message-input__content-editor"
                contenteditable="true"
                data-placeholder="Type message here"
              />
              <div
                class="ps__rail-x"
                style="left: 0px; top: 0px;"
              >
                <div
                  class="ps__thumb-x"
                  style="left: 0px; width: 0px;"
                  tabindex="0"
                />
              </div>
              <div
                class="ps__rail-y"
                style="top: 0px; left: 0px;"
              >
                <div
                  class="ps__thumb-y"
                  style="top: 0px; height: 0px;"
                  tabindex="0"
                />
              </div>
            </div>
          </div>
          <div
            class="cs-message-input__tools"
          >
            <button
              class="cs-button cs-button--send"
              disabled=""
            >
              <svg
                aria-hidden="true"
                class="svg-inline--fa fa-paper-plane fa-w-16 "
                data-icon="paper-plane"
                data-prefix="fas"
                focusable="false"
                role="img"
                viewBox="0 0 512 512"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M476 3.2L12.5 270.6c-18.1 10.4-15.8 35.6 2.2 43.2L121 358.4l287.3-253.2c5.5-4.9 13.3 2.6 8.6 8.3L176 407v80.5c0 23.6 28.5 32.9 42.5 15.8L282 426l124.6 52.2c14.2 6 30.4-2.9 33-18.2l72-432C515 7.8 493.3-6.8 476 3.2z"
                  fill="currentColor"
                />
              </svg>
            </button>
          </div>
        </div>
      </div>
    </div>
  </main>
</div>
`;

exports[`Dashboard renders correctly and fetches all users 1`] = `
<div>
  <main
    class="relative pr-[1px] h-full"
  >
    <nav
      class="flex justify-center gap-5 p-2 absolute z-50 right-2"
    >
      <a
        class="btn btn-sm btn-secondary btn-outline"
        href="/1"
      >
        Dashboard
      </a>
      <button
        class="btn btn-sm btn-accent"
      >
        Logout
      </button>
    </nav>
    <div
      class="cs-main-container h-screen min-h-[700px]"
    >
      <div
        class="cs-sidebar cs-sidebar--left"
      >
        <div
          class="flex justify-center items-center bg-slate-200 h-20"
        >
          <p>
            User1
          </p>
        </div>
        <div
          class="cs-conversation-list"
        >
          <div
            class="scrollbar-container ps"
          >
            <ul>
              <li>
                <div
                  class="cs-conversation  my-2"
                >
                  <div
                    class="cs-conversation__content"
                  >
                    <div
                      class="cs-conversation__name"
                    >
                      Test User2
                    </div>
                    <div
                      class="cs-conversation__info"
                    />
                  </div>
                </div>
              </li>
            </ul>
            <div
              class="ps__rail-x"
              style="left: 0px; top: 0px;"
            >
              <div
                class="ps__thumb-x"
                style="left: 0px; width: 0px;"
                tabindex="0"
              />
            </div>
            <div
              class="ps__rail-y"
              style="top: 0px; left: 0px;"
            >
              <div
                class="ps__thumb-y"
                style="top: 0px; height: 0px;"
                tabindex="0"
              />
            </div>
          </div>
        </div>
      </div>
      <div
        class="w-full items-center flex justify-center"
      >
        <div>
          Select someone to chat with
        </div>
      </div>
    </div>
  </main>
</div>
`;
